flink-runtime/src/test/java/org/apache/flink/runtime/testutils/InMemoryReporter.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.testutils;
+
+import org.apache.flink.annotation.Experimental;
+import org.apache.flink.metrics.LogicalScopeProvider;
+import org.apache.flink.metrics.Metric;
+import org.apache.flink.metrics.MetricConfig;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.metrics.groups.OperatorMetricGroup;
+import org.apache.flink.metrics.reporter.MetricReporter;
+import org.apache.flink.metrics.reporter.MetricReporterFactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.AbstractMap.SimpleEntry;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * A {@link MetricReporter} implementation that makes all reported metrics available for tests.
+ * Metrics remain registered even after the job finishes for easier assertions.
+ *
+ * <p>This class can only be accessed through {@link InMemoryReporterRule#getReporter()} to ensure
+ * that test cases are properly isolated.
+ */
+@Experimental
+public class InMemoryReporter implements MetricReporter {
```

Comment:
I think this class will be accessed from different threads but is not thread safe. This is a problem as seen in https://issues.apache.org/jira/browse/FLINK-23785.

Reply:
`MetricReporter`s are usually synchronized externally but I used it in an unintended way, so you are right it needs synchronization but for more elaborate reason. And it's not solving the whole issue, see the ticket.